### PR TITLE
Master lead enrichment fle

### PR DESCRIPTION
**Pull request by fleodoo**
Purpose:
Allow the user to enrich a lead based on the domain of the email adress of the client.
Following fields will be populated if found using clearbit and if they are not yet populated:
        - description
        - partner_name
        - reveal_id
        - website
        - street, street2, zip, city, country_id, state_id
        - phone (using the first number found in clearbit)
        - mobile (using the second number found in clearbit, else the same as phone)
A message will also be logged in the chatter.
If no data is found using clearbit no credit is used and a message will be logged,
If the user has no credits he will be redirected to buy credits for the service.
It is possible to enrich using a new secundary button on in the header of a lead,
or using the 'enrich' action in listview.

This commit includes:
	- a new botton in leads
	- a new action in listview
	- a new message tempalate to log when no data is found in clearbit
	- additions to crm_lead to be able to enrich

TaskId: 1931856

### CODE REVIEW

File: addons/crm_iap_lead/models/crm_lead.py [hunk 1]
```diff
@@ -1,11 +1,113 @@
...
+            messages_to_post = self._create_message_data(company_data)
+            self.message_post_with_view('crm_iap_lead.lead_message_template', values=messages_to_post, subtype_id=self.env.ref('mail.mt_note').id)
+        else:
+            self.message_post_with_view('crm_iap_lead.lead_message_not_found', subtype_id=self.env.ref('mail.mt_note').id)
+
+    @api.multi
+    def _enrich_leads_from_response(self, data):
+        """ This method will get the response from the service and enrich the lead accordingly """
+        self.ensure_one()
+        if not self.description:
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: Create a dict with fields to update and call `self.write(values)` at the end.
```
vals = {
    'description': self.description or data['description'],
    'partner_name': self.partner_name or data['name'],
    ...
}
if not self.street ... : 
    vals.update({
        'street': data['street_name'],
        ...
    })
self.write(vals)
```

File: addons/crm_iap_lead/models/crm_lead.py [hunk 2]
```diff
@@ -1,11 +1,113 @@
...
 class Lead(models.Model):
     _inherit = 'crm.lead'
 
     reveal_id = fields.Char(string='Reveal ID', index=True)
     lead_mining_request_id = fields.Many2one('crm.iap.lead.mining.request', string='Lead Mining Request', index=True)
+
+    @api.multi
+    def lead_enrich_mail(self):
+        self.ensure_one()
+        banned_domains =["gmail.com", "yahoo.com", "hotmail.com", "aol.com"]
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: Create a new model to be able to add or remove domains

File: addons/crm_iap_lead/models/crm_lead.py [hunk 3]
```diff
@@ -1,11 +1,113 @@
...
 
 class Lead(models.Model):
     _inherit = 'crm.lead'
 
     reveal_id = fields.Char(string='Reveal ID', index=True)
     lead_mining_request_id = fields.Many2one('crm.iap.lead.mining.request', string='Lead Mining Request', index=True)
+
+    @api.multi
+    def lead_enrich_mail(self):
+        self.ensure_one()
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: remove, and use `for record in self:`

File: addons/crm_iap_lead/models/crm_lead.py [hunk 4]
```diff
@@ -1,11 +1,141 @@
...
+            self.message_post_with_view('crm_iap_lead.lead_message_not_found', subtype_id=self.env.ref('mail.mt_note').id)
+
+    @api.multi
+    def _enrich_leads_from_response(self, data_clearbit):
+        """ This method will get the response from the service and enrich the lead accordingly """
+        for lead_id, data in data_clearbit.items():
+            record = self.browse(int(lead_id))
+            if record.exists():
+                if data:
+                    street = False
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: useless variable, in your dict, do
`'street': record.street or data['street_name']`

File: addons/crm_iap_lead/models/crm_lead.py [hunk 5]
```diff
@@ -1,11 +1,141 @@
...
+                    country_id = False
+                    state_id = False
+
+                    if not(record.street or record.street2 or record.zip or record.city or record.state_id or record.country_id):
+                        street = data["street_number"]
+                        street2 = data["street_name"]
+                        _zip = data["postal_code"]
+                        city = data["city"]
+                        country = record.env['res.country'].search([('code', '=', data["country_code"])])
+                        country_id = country.id
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: country could be none, be careful

File: addons/crm_iap_lead/models/crm_lead.py [hunk 6]
```diff
@@ -1,11 +1,141 @@
...
+                        state = record.env['res.country.state'].search([('code', '=', data["state_code"]),('country_id', '=', country_id)])
+                        state_id = state.id
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: Same here

File: addons/crm_iap_lead/models/crm_lead.py [hunk 7]
```diff
@@ -1,11 +1,141 @@
...
+                        'country_id': record.country_id.id or country_id,
+                        'state_id': record.state_id.id or state_id,
+                    }
+                    record.write(vals)
+                    record._send_message(data)
+                else:
+                    record._send_message()
+
+
+class banned_domain(models.Model):
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: BannedDomain

File: addons/iap/static/src/js/crash_manager.js [hunk 8]
```diff
@@ -20,7 +20,7 @@ CrashManager.include({
      */
     _getButtonMessage: function (isTrial){
         var isEnterprise = _.last(odoo.session_info.server_version_info) === 'e';
-        return isTrial && isEnterprise ? _t('Start a Trial at Odoo') : _t('Buy credits at Odoo');
+        return isTrial && isEnterprise ? _t('Start a Trial at Odoo') : _t('Buy credits');
```
Comment (pro-odoo) [inline addons/iap/static/src/js/crash_manager.js]: Is it intended ?

File: addons/crm_iap_lead/security/ir.model.access.csv [hunk 9]
```diff
@@ -5,3 +5,4 @@ access_crm_iap_lead_seniority,access_crm_iap_lead_seniority,model_crm_iap_lead_s
 access_crm_iap_lead_technology,access_crm_iap_lead_technology,model_crm_iap_lead_technology,sales_team.group_sale_manager,1,1,1,1
 access_crm_iap_lead_mining_request,access_crm_iap_lead_mining_request,model_crm_iap_lead_mining_request,sales_team.group_sale_manager,1,1,1,1
 access_crm_iap_lead_helpers,access_crm_iap_lead_helpers,model_crm_iap_lead_helpers,,0,0,0,0
+access_crm_iap_lead_banned_domain,access_crm_iap_lead_banned_domain,model_crm_iap_lead_banned_domain,base.group_user,1,0,0,0
```
Comment (pro-odoo) [inline addons/crm_iap_lead/security/ir.model.access.csv]: No one can create banned domain ?

File: addons/crm_iap_lead/data/crm.iap.banned.domains.xml [hunk 10]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
```
Comment (pro-odoo) [inline addons/crm_iap_lead/data/crm.iap.banned.domains.xml]: Only four ?

Comment (pro-odoo) [reply addons/crm_iap_lead/data/crm.iap.banned.domains.xml]: In addition, to it in csv

File: addons/iap/static/src/js/crash_manager.js (hunk 8 above)
Comment (fleodoo) [reply addons/iap/static/src/js/crash_manager.js]: this change was asked by aju

File: addons/crm_iap_lead/models/crm_lead.py [hunk 11]
```diff
@@ -1,11 +1,125 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import fields, models
+from odoo import fields, models, api, _
+from odoo.exceptions import UserError
+from odoo.addons.iap import jsonrpc,InsufficientCreditError
 
+# DEFAULT_ENDPOINT = 'https://iap-services.odoo.com'
+DEFAULT_ENDPOINT = 'http://iap.vcap.me:8269'
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: Remove it ;-)

File: addons/crm_iap_lead/models/crm_lead.py [hunk 12]
```diff
@@ -1,11 +1,125 @@
...
     lead_mining_request_id = fields.Many2one('crm.iap.lead.mining.request', string='Lead Mining Request', index=True)
+
+    @api.multi
+    def lead_enrich_mail(self):
+        request = {}
+        for record in self:
+            if record.probability == 0:
+                continue
+            email_domain = record.partner_address_email or record.email_from
+            print(email_domain)
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: Remove it

File: addons/crm_iap_lead/models/crm_lead.py [hunk 13]
```diff
@@ -1,11 +1,125 @@
...
+        self._enrich_leads_from_response(response_clearbit)
+
+    @api.multi
+    def lead_enrich_mail_server_action(self):
+        # This function has as purpose to differentiate the way the errors are handeled if the code is called by
+        # A server action or using the normal flow. This because InsufficientCreditError are not handeled in the server actions.
+        try:
+            return self.lead_enrich_mail()
+        except InsufficientCreditError:
+            raise UserError("You do not have enough credits to perform this action.")
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: This message is not translatable... Use _()

File: addons/crm_iap_lead/models/crm_lead.py [hunk 14]
```diff
@@ -1,11 +1,125 @@
...
+
+        response_clearbit = record._make_request(request)
+        self._enrich_leads_from_response(response_clearbit)
+
+    @api.multi
+    def lead_enrich_mail_server_action(self):
+        # This function has as purpose to differentiate the way the errors are handeled if the code is called by
+        # A server action or using the normal flow. This because InsufficientCreditError are not handeled in the server actions.
+        try:
+            return self.lead_enrich_mail()
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: this method does not return anything, simply do `self.lead_enrich_mail()`

File: addons/crm_iap_lead/models/crm_lead.py [hunk 15]
```diff
@@ -1,11 +1,125 @@
...
+        dbuuid = self.env['ir.config_parameter'].sudo().get_param('database.uuid')
+        endpoint = self.env['ir.config_parameter'].sudo().get_param('reveal.endpoint', DEFAULT_ENDPOINT) + '/iap/clearbit/1/lead_enrichment_email'
+        params = {
+            'account_token': reveal_account.account_token,
+            'dbuuid': dbuuid,
+            'domains': domains,
+        }
+        return jsonrpc(endpoint, params=params, timeout=300)
+
+    def _create_message_data(self, company_data):
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: api.model

File: addons/crm_iap_lead/models/crm_lead.py [hunk 16]
```diff
@@ -1,11 +1,125 @@
...
+        }
+        timezone = company_data['timezone']
+        if timezone:
+            log_data.update({
+                'timezone': timezone.replace('_', ' ').title(),
+                'timezone_url': company_data['timezone_url'],
+            })
+        return log_data
+
+    def _send_message(self, company_data=False):
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: api.model

File: addons/crm_iap_lead/models/crm_lead.py [hunk 17]
```diff
@@ -1,11 +1,125 @@
...
+            self.message_post_with_view('crm_iap_lead.lead_message_not_found', subtype_id=self.env.ref('mail.mt_note').id)
+
+    @api.model
+    def _enrich_leads_from_response(self, data_clearbit):
+        """ This method will get the response from the service and enrich the lead accordingly """
+        for lead_id, data in data_clearbit.items():
+            record = self.browse(int(lead_id))
+            if record.exists():
+                if data:
+                    print(data)
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: You need to reread yourself before mark a task as ready, it's the third print in the code ;-)

File: addons/crm_iap_lead/models/crm_lead.py [hunk 18]
```diff
@@ -1,11 +1,125 @@
...
+                    street = False
+                    street2 = False
+                    zip_code = False
+                    city = False
+                    country_id = False
+                    state_id = False
+
+                    if not(record.street or record.street2 or record.zip or record.city or record.state_id or record.country_id):
+                        street = ' '.join(filter(False, (data["street_name"], data["street_number"])))
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: https://docs.python.org/3/library/functions.html#filter
False is not a function

File: addons/crm_iap_lead/models/crm_lead.py [hunk 19]
```diff
@@ -1,11 +1,125 @@
...
+                        street2 = data["location"]
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: Remove it, the location is the whole string of the address, not the street2

File: addons/crm_iap_lead/models/crm_lead.py [hunk 20]
```diff
@@ -1,11 +1,125 @@
...
+                        zip_code = data["postal_code"]
+                        city = data["city"]
+                        country = record.env['res.country'].search([('code', '=', data["country_code"])])
+                        if country:
+                            country_id = country.id
+                            state = record.env['res.country.state'].search([('code', '=', data["state_code"]),('country_id', '=', country_id)])
+                            if state:
+                                state_id = state.id
+                    vals = {
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: You do not reuse vals, so do `record.write({...})`

File: addons/crm_iap_lead/models/crm_lead.py [hunk 21]
```diff
@@ -1,11 +1,125 @@
...
+                        'description': record.description or data['description'],
+                        'partner_name': record.partner_name or data['name'],
+                        'reveal_id': record.reveal_id or data['clearbit_id'],
+                        'website': record.website or ('https://www.%s' % data['domain'] if data['domain'] else False),
+                        'phone': record.phone or (data["phone_numbers"][0] if (len(data["phone_numbers"])>0) else False),
+                        'mobile': record.mobile or (data["phone_numbers"][1] if (len(data["phone_numbers"])>1) else False),
+                        'street': record.street or street,
+                        'street2': record.street2 or street2,
```
Comment (pro-odoo) [inline addons/crm_iap_lead/models/crm_lead.py]: as you remove street2, remove this entry

File: addons/crm_iap_lead/views/crm_iap_lead_enrichment_email.xml [hunk 22]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<odoo>
+    <data>
+        <record model="ir.ui.view" id="crm_enrich_mail">
+                <field name="model">crm.lead</field>
```
Comment (pro-odoo) [inline addons/crm_iap_lead/views/crm_iap_lead_enrichment_email.xml]: add a name to the view

File: addons/iap/static/src/js/crash_manager.js (hunk 8 above)
Comment (pro-odoo) [reply addons/iap/static/src/js/crash_manager.js]: Ok

**Comment by pro-odoo** [comment]
https://github.com/odoo/odoo/pull/33812

**Pull request closed without merging**